Changes in module 'nf-core/samtools/fastq'
--- modules/nf-core/samtools/fastq/main.nf
+++ modules/nf-core/samtools/fastq/main.nf
@@ -9,13 +9,9 @@
 
     input:
     tuple val(meta), path(input)
-    val(interleave)
 
     output:
-    tuple val(meta), path("*_{1,2}.fastq.gz")      , optional:true, emit: fastq
-    tuple val(meta), path("*_interleaved.fastq.gz"), optional:true, emit: interleaved
-    tuple val(meta), path("*_singleton.fastq.gz")  , optional:true, emit: singleton
-    tuple val(meta), path("*_other.fastq.gz")      , optional:true, emit: other
+    tuple val(meta), path("*.fastq.gz")      , optional:true, emit: fastq
     path  "versions.yml"                           , emit: versions
 
     when:
@@ -24,17 +20,13 @@
     script:
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
-    def output = ( interleave && ! meta.single_end ) ? "> ${prefix}_interleaved.fastq.gz" :
-        meta.single_end ? "-1 ${prefix}_1.fastq.gz -s ${prefix}_singleton.fastq.gz" :
-        "-1 ${prefix}_1.fastq.gz -2 ${prefix}_2.fastq.gz -s ${prefix}_singleton.fastq.gz"
     """
     samtools \\
         fastq \\
         $args \\
         --threads ${task.cpus-1} \\
-        -0 ${prefix}_other.fastq.gz \\
         $input \\
-        $output
+        > ${prefix}.fastq.gz
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
